R/tbl_cross.R

```diff
@@ -0,0 +1,163 @@
+#' \if{html}{\figure{tbl_cross_ex.png}{options: width=50\%}}
+#'
+tbl_cross <- function(data,
+                      row = NULL,
+                      col = NULL,
+                      label = NULL,
+                      statistic = NULL,
+                      percent = c("none", "column", "row", "cell"),
+                      missing = c("ifany", "always", "no"),
+                      missing_text = "Unknown") {
+  row_str <- var_input_to_string(
+    data = data,
+    select_input = !!rlang::enquo(row),
+    arg_name = "row_str",
+    select_single = TRUE
+  )
+
+  col_str <- var_input_to_string(
+    data = data,
+    select_input = !!rlang::enquo(col),
+    arg_name = "col_str",
+    select_single = TRUE
+  )
+
+  # matching arguments ---------------------------------------------------------
+  missing <- match.arg(missing)
+  percent <- match.arg(percent)
+
+
+  # if no x and y provided, default to first two col_strumns of data -----------
+  if (is.null(row_str) & is.null(col_str)) {
+    row_str <- names(data[, 1])
+    col_str <- names(data[, 2])
+  }
+
+  row_str <- row_str %||% names(select(data, -tidyselect::all_of(col_str)))[1]
+  col_str <- col_str %||% names(select(data, -tidyselect::all_of(row_str)))[1]
+
+  # create new dummy col for tabulating column totals in cross table
+  data <- data %>%
+    dplyr::select(tidyselect::all_of(c(row_str, col_str))) %>%
+    mutate(Total = 1)
+
+  # get col label (for tab spanner)--------------------------------------------
+  label <- tidyselect_to_list(data, label) %||% NULL
+
+  col_label <- label[[col_str]] %||%
+    attr(data[[col_str]], "label") %||% col_str
+
+  # statistic argument ---------------------------------------------------------
+
+  # if no user-defined stat, default to {n} if percent is "none"
+  statistic <- statistic %||% ifelse(percent == "none", "{n}", "{n} ({p}%)")
+  statistic_list <- rep(statistic, 3) %>%
+    as.list() %>%
+    set_names(names(data))
+
+  # omit missing data, or factorize missing level ------------------------------
+  lbls <- purrr::map(data, ~ attr(.x, "label"))
+
+  data <- data %>%
+    mutate_at(vars(row_str, col_str), ~ as.factor(.x)) %>%
+    mutate_at(
+      vars(row_str, col_str),
+      ~ switch(
+        missing,
+        "no" = .,
+        "ifany" = forcats::fct_explicit_na(., missing_text),
+        "always" = forcats::fct_explicit_na(., missing_text) %>%
+          forcats::fct_expand(missing_text)
+      )
+    )
+
+  if (missing == "no") {
+    data <- stats::na.omit(data)
+
+    message(glue(
+      "{sum(is.na(data))} observations missing data have been removed."
+    ))
+  }
+
+  # re-applying labels
+  for (i in names(lbls)) {
+    attr(data[[i]], "label") <- lbls[[i]]
+  }
+
+
+  # create main table ----------------------------------------------------------
+  x <- data %>%
+    tbl_summary(
+      by = col_str,
+      statistic = statistic_list,
+      percent = switch((percent == "none") + 1, percent, NULL),
```

Comment:
don't you wish there was a more elegant way of doing this?!?

what do you think of this instead?
`percent = switch(percent != "none", percent)`

Reply:
I'm still getting the hang of `switch()`, but I think that would work! Does that mean if percent does in fact equal "none, it will return NULL? When I try it alone it doesn't explicitly return `NULL` but I will test it in context!